pkg/node/node.go

```diff
@@ -101,105 +103,261 @@ func createAddressFilter(
+			keepAddress.String(),
+			attemptCounter,
+		)
+
+		// If we are re-attempting the key generation, pre-parameters in the box
+		// could be destroyed because they were shared with other members.
+		// In this case, we need to re-generate them.
+		if preParamsBox.IsEmpty() {
+			logger.Infof("Getting new key-generation pre-parameters from the pool")
+			preParamsBox = params.NewBox(n.tssParamsPool.get())
+			logger.Infof("Received new key-generation parameters from the pool")
+		}
+
+		// Global timeout for generating a signer exceeded.
+		// We are giving up and leaving this function.
+		if ctx.Err() != nil {
+			return nil, fmt.Errorf("key generation timeout exceeded")
+		}
+
+		// Announce signer presence. Other members of the keep need to receive
+		// the public key of this members. This member, need to receive public
+		// keys of all other members. Up to this point, only addresses from
+		// signer selection protocol are known.
+		//
+		// If signer announcement fails, we retry from the beginning.
+		memberIDs, err := n.AnnounceSignerPresence(
+			ctx,
+			operatorPublicKey,
+			keepAddress,
+			members,
+		)
+		if err != nil {
+			logger.Errorf("failed to announce signer presence: [%v]", err)
+			continue
+		}
+
+		// Generate threshold signer by generating threshold key with all other
+		// keep members.
+		//
+		// If threshold key generation fails, we retry from the beginning.
+		signer, err := tss.GenerateThresholdSigner(
+			ctx,
+			keepAddress.Hex(),
+			memberID,
+			memberIDs,
+			uint(len(memberIDs)-1),
+			n.networkProvider,
+			preParamsBox,
+		)
+		if err != nil {
+			logger.Errorf("failed to generate threshold signer: [%v]", err)
+			continue
+		}
+
+		// Serialize and publish public key to the keep.
+		//
+		// We don't retry in case of an error although the specific chain
+		// implementation may implement its own retry policy. This action
+		// should never fail and if it failed, something terrible happened.
+		publicKey, err := eth.SerializePublicKey(signer.PublicKey())
+		if err != nil {
+			return nil, fmt.Errorf("failed to serialize public key: [%v]", err)
+		}
+		err = n.publishSignerPublicKey(ctx, keepAddress, publicKey)
+		if err != nil {
+			return nil, err
+		}
+
+		return signer, nil // key generation succeeded.
 	}
+}
 
+func (n *Node) publishSignerPublicKey(
+	ctx context.Context,
+	keepAddress common.Address,
+	publicKey [64]byte,
+) error {
 	logger.Debugf(
-		"generated threshold signer with public key: [%x]",
-		signer.PublicKey().Marshal(),
+		"submitting public key to the keep [%s]: [%x]",
+		keepAddress.String(),
+		publicKey,
 	)
 
-	// Publish signer's public key on ethereum blockchain in a specific keep
-	// contract.
-	serializedPublicKey, err := eth.SerializePublicKey(signer.PublicKey())
-	if err != nil {
-		return nil, fmt.Errorf("failed to serialize public key: [%v]", err)
-	}
-
 	monitoringAbort := make(chan interface{})
 	go n.monitorKeepPublicKeySubmission(ctx, monitoringAbort, keepAddress)
```

Comment:
What will happen when one of the members fails on key submission?
Maybe we should block on the monitoring and wait for the successful key establishment notification from the keep and restart generation in case of timeout? Or just retry submission as we do for signatures?

Reply:
Inside `ethereum.go` we have a retry for a situation when the public key submission fails. If the public key submission fails although it was retried multiple times, there is really nothing we can do. Some other member could submit their key so we can't re-run the protocol one more time. Implementing an on-chain cleanup for failed key submission can be an attack vector: I failed to manipulate threshold key generation and I don't publish the result so that I have yet one chance.

Reply:
Since publishing the key is so important and we cannot retry the whole process maybe we could give it more chances to submit the tx? Right now we have just 4 retries every 250 ms:
https://github.com/keep-network/keep-ecdsa/blob/3f2d73fe07e9f897ff0fb5e138ac7047a590aea0/pkg/chain/ethereum/ethereum.go#L225-L241

While testing it with Infura I observed a couple of times that this time frame was not enough to submit the transaction successfully.

Reply:
I'll increase it to 10 retries, every 12 seconds, cc @dimpar 
It gives a total of 2 minutes.

Reply:
See https://github.com/keep-network/keep-ecdsa/pull/357/commits/f23eb95888d54ccaaab6576a078d8d6a96121f36